### PR TITLE
Add support for -XX:[+|-]HandleSIGABRT

### PR DESCRIPTION
**Pull request by babsingh**
**a) Add a tracepoint to output an error for incompatible cmdline options**

The new tracepoint takes the two incompatible cmdline options as string
inputs.

**b) Add macros to support `-XX:[+|-]HandleSIGABRT`**

**c) Parse the new cmdline option: `-XX:[+|-]HandleSIGABRT`**

```
1. default                               -> JVM abort handler enabled
2. -XX:+HandleSIGABRT                    -> JVM abort handler enabled
3. -XX:-HandleSIGABRT                    -> JVM abort handler disabled
4. -XX:-HandleSIGABRT -XX:+HandleSIGABRT -> JVM abort handler enabled
5. -XX:+HandleSIGABRT -XX:-HandleSIGABRT -> JVM abort handler disabled
6. -XX:+HandleSIGABRT -Xrs               -> not allowed, error thrown
7. ... -XX:+HandleSIGABRT -Xrs:sync      -> not allowed, error thrown
8. ... -XX:-HandleSIGABRT -Xrs...        -> JVM abort handler disabled
```
`setSignalOptions` returns an error code if incompatible command line
options are specified.

`initializeJavaVM` terminates with `JNI_ERR` if `setSignalOptions` returns an
error code.

**d) Install the JVM abort handler based on `-XX:[+|-]HandleSIGABRT`**

**e) Add tests for `-XX:[+|-]HandleSIGABRT`**

- Copied `functional/cmdLineTests/gptest` to create `sigabrtHandlingTest`
- `class GPTest` updated to output `"Invoking abort!"`
- `sigabrtHandlingTest.xml` contains testcases for `-XX:[+|-]HandleSIGABRT`

Related: https://github.com/eclipse/openj9/issues/8735, https://github.com/eclipse/openj9-docs/pull/564

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
**Notes:**
- These changes have been tested on Windows, zOS, AIX and Linux.
- Targeted for the 0.21 release.
- Command line option implemented as per the discussion in https://github.com/eclipse/openj9/issues/8735.

**Reviewers:**
- @llxia for the test commit, `Add tests for -XX:[+|-]HandleSIGABRT`.
- @DanHeidinga for the VM changes.

File: runtime/vm/jvminit.c [hunk 1]
```diff
@@ -6519,6 +6526,23 @@ setSignalOptions(J9JavaVM* vm)
...
+	if (argIndex2 > argIndex) {
+		/* Enable the JVM abort handler since -XX:+HandleSIGABRT is seen last. */
+		if (OMR_ARE_ALL_BITS_SET(vm->sigFlags, J9_SIG_XRS_SYNC)) {
+			/* Throw error message if both -XX:+HandleSIGABRT and -Xrs/-Xrs:sync are supplied. */
+			j9nls_printf(portLibrary, J9NLS_ERROR, J9NLS_VM_INCOMPATIBLE_CMDLINE_OPTIONS_ERROR, VMOPT_XXHANDLESIGABRT, VMOPT_XRS);
+			return -1;
+		}
+	} else if (argIndex > argIndex2) {
+		/* Disable the JVM abort handler since -XX:-HandleSIGABRT is seen last. */
+		vm->sigFlags |= J9_SIG_NO_SIG_ABRT;
```
Comment (DanHeidinga) [inline runtime/vm/jvminit.c:6541]: Does the omr signal code need to be notified about this - ie by `j9sig_set_options`?

Comment (babsingh) [reply runtime/vm/jvminit.c:6541]: No, the OMR signal code doesn't need to be notified. The VM handles `SIGABRT` from [rasdump/dmpsup.c](https://github.com/eclipse/openj9/blob/master/runtime/rasdump/dmpsup.c). The option is evaluated on the VM side (`dmpsup.c`) before any of the basic OMR signal functions are invoked. So, OMR doesn't need to be aware of this option.

In OMR, `SIGABRT` is categorized as an `async` signal. A Java user can potentially register a **Java signal handler** for `SIGABRT`: `sun.misc.Signal.handle(...)` -> `JVM_RegisterSignal(...)` -> `omrsig_set_async_signal_handler(...)`.  We don't want to influence this code-path with the new option. This option only enables or disables the VM signal handler for `SIGABRT`, which is maintained in `dmpsup.c`.

File: runtime/rasdump/dmpsup.c [hunk 2]
```diff
@@ -1095,14 +1095,12 @@ pushDumpFacade(J9JavaVM *vm)
 		/* Install new facade */
 		vm->j9rasDumpFunctions = (J9RASdumpFunctions *)queue;
 
-		/* Note that we need to do the same check in popDumpFacade()
-		 *
-		 * According to the port library SIGABRT (J9PORT_SIG_FLAG_SIGABRT) is an ASYNC signal, so you'd think we should be checking for J9PORT_SIG_OPTIONS_REDUCED_SIGNALS_ASYNCHRONOUS here.
-		 *  However we want to disable the abortHandler when either of -Xrs / -Xrs:sync are present, so check for J9PORT_SIG_OPTIONS_REDUCED_SIGNALS_SYNCHRONOUS instead
+		/* Note that we need to do the same check in popDumpFacade().
+		 * Do not install the abort handler if -Xrs, -Xrs:sync or -XX:-HandleSIGABRT are present.
          */
```
Comment (DanHeidinga) [inline runtime/rasdump/dmpsup.c:1100]: Minor thing but can you fix the formatting of this comment?  The closing `*/` doesn't line up correctly (and didn't in the original code either)

File: runtime/rasdump/dmpsup.c [hunk 3]
```diff
@@ -1121,12 +1119,10 @@ popDumpFacade(J9JavaVM *vm)
 	if ( FIND_DUMP_QUEUE(vm, queue) )
 	{
 		PORT_ACCESS_FROM_JAVAVM(vm);
-		/* Note that we need to do the same check in pushDumpFacade()
-		 *
-		 * According to the port library SIGABRT (J9PORT_SIG_FLAG_SIGABRT) is an ASYNC signal, so you'd think we should be checking for J9PORT_SIG_OPTIONS_REDUCED_SIGNALS_ASYNCHRONOUS here.
-		 *  However we want to disable the abortHandler when either of -Xrs / -Xrs:sync are present, so check for J9PORT_SIG_OPTIONS_REDUCED_SIGNALS_SYNCHRONOUS instead
+		/* Note that we need to do the same check in pushDumpFacade().
+		 * Do not install the abort handler if -Xrs, -Xrs:sync or -XX:-HandleSIGABRT are present.
          */
```
Comment (DanHeidinga) [inline runtime/rasdump/dmpsup.c:1124]: same nitpick here

**Comment by DanHeidinga** [comment]
I'm OK with this change once the conflicts are resolved.

@llxia can you review the tests?

**Comment by babsingh** [comment]
The minor nitpicks and conflicts have been fixed.

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		</levels>
+		<groups>
+			<group>functional</group>
+		</groups>
+		<types>
+			<type>native</type>
+		</types>
+		<subsets>
+			<subset>9</subset>
+			<subset>11+</subset>
```
Comment (llxia) [inline test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:92]: we can just use `<subset>9+</subset>`

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+<playlist xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../TKG/playlist.xsd">
+	<test>
+		<testCaseName>cmdLineTest_sigabrtHandlingTest</testCaseName>
+		<variations>
+			<variation>NoOptions</variation>
+		</variations>
```
Comment (llxia) [inline test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:30]: The default is NoOptions. We can remove line 28-30. Same for the test below.

File: test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  [2] http://openjdk.java.net/legal/assembly-exception.html
+
+  SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+<!DOCTYPE suite SYSTEM "cmdlinetester.dtd">
+
+<suite id="SIGABRT Handling Tests" timeout="2400">
+ <variable name="CP" value="-cp $Q$$RESJAR$$Q$ VMBench.GPTests.GPTest" />
+ <variable name="THREADED" value=" " />
```
Comment (llxia) [inline test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml:29]: Do we need `THREADED `?

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+		<command>$(ADD_JVM_LIB_DIR_TO_LIBPATH) \
+	$(JAVA_COMMAND) $(JVM_OPTIONS) \
+	-DRESJAR=$(CMDLINETESTER_RESJAR) \
+	-DEXE=$(SQ)$(JAVA_COMMAND) $(JVM_OPTIONS)$(SQ) \
+	-DJAVA_EXECUTABLE_DIR=$(Q)$(TEST_JDK_HOME)$(D)bin$(Q) \
```
Comment (llxia) [inline test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:35]: Did `JAVA_EXECUTABLE_DIR`, `UTILS_DIR`, `GLAUNCH`, `JVMLIBPATH`, `FIBTARGET` and `JVM_OPTIONS` get used?

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		<platformRequirements></platformRequirements>
+		<levels>
+			<level>sanity</level>
+		</levels>
+		<groups>
+			<group>functional</group>
+		</groups>
+		<types>
+			<type>native</type>
+		</types>
```
Comment (llxia) [inline test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:53]: This test is a java test to me. If so, we do not need to add `native `.  And we may also be able to remove `ADD_JVM_LIB_DIR_TO_LIBPATH `

File: test/functional/cmdLineTests/sigabrtHandlingTest/build.xml [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  Exception [1] and GNU General Public License, version 2 with the
+  OpenJDK Assembly Exception [2].
+
+  [1] https://www.gnu.org/software/classpath/license.html
+  [2] http://openjdk.java.net/legal/assembly-exception.html
+
+  SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+<project name="cmdLineTests" default="build" basedir=".">
```
Comment (llxia) [inline test/functional/cmdLineTests/sigabrtHandlingTest/build.xml:25]: The project name should be sigabrtHandlingTest

File: test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+ </test>
+ 
+ <test id="-XX:-HandleSIGABRT -Xrs">
+  <command>$EXE$ -XX:-HandleSIGABRT -Xrs $CP$ $THREADED$ abort</command>
+  <output type="success" regex="no">Invoking abort!</output>
+  <output type="failure" regex="no">Processing dump event "abort"</output>
+ </test>
+ 
+  <test id="-XX:-HandleSIGABRT -Xrs:sync">
+  <command>$EXE$ -XX:-HandleSIGABRT -Xrs:sync $CP$ $THREADED$ abort</command>
```
Comment (llxia) [inline test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml:88]: Please fix the identification.

Comment (pshipton) [reply test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml:88]: indentation...

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml (hunk 4 above)
Comment (DanHeidinga) [reply test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:92]: Thanks for calling this out.  Java 9 is dead.  There's no reason to reference anything other than 8, 11 or non-lts releases > 11.

If `9+` is fine, then please use `11+`

File: test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml (hunk 10 above)
Comment (llxia) [reply test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml:88]: yep, I meant indentation. :)

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml (hunk 8 above)
Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:53]: I need to set env. var. `NATIVE_TEST_LIBS` before running this test locally. Is type `native` responsible for setting `NATIVE_TEST_LIBS` in the builds?

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml (hunk 7 above)
Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:35]: Should they only be used in `sigabrtHandlingTest.xml`?
Can I remove them if they are not seen in `sigabrtHandlingTest.xml`?

Comment (llxia) [reply test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:35]: yes, I think so

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml (hunk 8 above)
Comment (llxia) [reply test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:53]: which native lib are you using?

Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:44]: `gptest`

**Comment by babsingh** [comment]
@llxia I have accounted for all your feedback in `Changes requested in the test code review` [commit]. If you are satisfied with these changes, I will squash it with the test commit.

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml (hunk 4 above)
Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml:76]: Removed: `<subset>9</subset>`

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml (hunk 5 above)
Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml]: Removed lines 28-30

File: test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml (hunk 6 above)
Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml]: Removed `THREADED`, not needed.

File: test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml (hunk 7 above)
Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/playlist.xml]: Removed all unused Java properties.

File: test/functional/cmdLineTests/sigabrtHandlingTest/build.xml (hunk 9 above)
Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/build.xml]: Fixed project name

File: test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml (hunk 10 above)
Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml]: Fixed indentation

**Comment by llxia** [comment]
Thanks @babsingh . Is there a difference between test for 8 and 11+ in playlist.xml? If not, then we can merge them into one.

**Comment by babsingh** [comment]
> Is there a difference between test for 8 and 11+ in playlist.xml?

Not anymore since `-DJVMLIBPATH` was removed from `playlist.xml`. Now, there is only one test in `playlist.xml` for both `8` and `11+`.

**Comment by DanHeidinga** [comment]
Jenkins test sanity xlinux jdk8

**Comment by DanHeidinga** [comment]
Jenkins test sanity win,osx jdk11

**Comment by DanHeidinga** [comment]
@pshipton Any further comments or concerns with this?  Otherwise it's ready to merge

**Comment by pshipton** [comment]
A number of the tests are producing system core files. I think they should be disabled. Having the other dump types, or just one of them, should be sufficient.

**Comment by babsingh** [comment]
The system core files have been disabled in the HEAD commit.

**Comment by pshipton** [comment]
@DanHeidinga I've approved it.

File: test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  [2] http://openjdk.java.net/legal/assembly-exception.html
+
+  SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+<!DOCTYPE suite SYSTEM "cmdlinetester.dtd">
+
+<suite id="SIGABRT Handling Tests" timeout="2400">
+ <variable name="CP" value="-cp $Q$$RESJAR$$Q$ VMBench.GPTests.GPTest" />
+ <variable name="DUMP" value="-Xdump:system:none" />
```
Comment (DanHeidinga) [inline test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml:29]: Can you add a comment to this to indicate why this has been added?

Comment (babsingh) [reply test/functional/cmdLineTests/sigabrtHandlingTest/sigabrtHandlingTest.xml]: A comment has been added to indicate why the `-Xdump` option is added.

**Comment by DanHeidinga** [comment]
Jenkins test sanity osx jdk11

**Comment by DanHeidinga** [comment]
Running one last set of builds to validate the test change and then this will be ready to merge

**Comment by babsingh** [comment]
[Build_JDK11_x86-64_mac_Personal](https://ci.eclipse.org/openj9/job/Build_JDK11_x86-64_mac_Personal/509/console) failed because of
```
09:30:59  JVMDUMP013I Processed dump event "systhrow", detail "java/lang/ClassCastException".
09:30:59  java.lang.ClassCastException: java.lang.String incompatible with java.nio.file.FileTreeWalker$Event
09:30:59  	at java.base/java.nio.file.Files$$Lambda$6/0000000000000000.test(Unknown Source)
09:30:59  	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
```
This failure is a known issue reported in https://github.com/eclipse/openj9/issues/8173. A restart should fix it.

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk11

**Comment by DanHeidinga** [comment]
launching zlinux as it tends to build faster

**Pull request merged**